src/evaluators/LogicalExpression.js

```diff
@@ -93,5 +93,9 @@ export default function(
   // return or throw completion
   if (completion instanceof AbruptCompletion) throw completion;
   invariant(completion instanceof NormalCompletion || completion instanceof Value || completion instanceof Reference);
+  if (lval instanceof Value && compl2 instanceof Value) {
+    // Reconstruct the logical op to help simplification and reasoning
+    completion = AbstractValue.createFromLogicalOp(realm, ast.operator, lval, compl2, ast.loc);
```

Comment:
Is this a temporary hack? What exactly goes wrong with the regular joining logic?

Reply:
The regular joining logic is not wrong, but constructs inconvenient expressions. You are quite right to wonder why make a special case of only the return value. That should get fixed too, but probably not in this request.

Reply:
What is the "inconvenient expressions"? Maybe it's all going to work out fine in the end, but I think it would be good to start painting a holistic picture of the algebra involved --- what expressions are being constructed, how does a normal formal look like that we can reliably and automatically compute, why is there going to be confluence in the end...

Reply:
Can you add a TODO with a comment that calls out the particular inconvenient expression that this is working around?

Reply:
I'll amplify the comment, but thinking about this some more, the join expressions for the side effects of the second argument are just fine as they are. So no TODO needed here. This change makes things the way I want to them to be.